adapters/adxcg/adxcg.go

```diff
@@ -0,0 +1,130 @@
+	return bidder, nil
+}
+
+// MediaOptiAdapter implements the Bidder interface.
+type AdxcgAdapter struct {
+	endpoint string
+}
+
+// MakeRequests prepares the HTTP requests which should be made to fetch bids.
+func (adapter *AdxcgAdapter) MakeRequests(
+	openRTBRequest *openrtb2.BidRequest,
+	reqInfo *adapters.ExtraRequestInfo,
+) (
+	requestsToBidder []*adapters.RequestData,
+	errs []error,
+) {
+	openRTBRequestJSON, err := json.Marshal(openRTBRequest)
+	if err != nil {
+		errs = append(errs, err)
+		return nil, errs
+	}
+
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	requestToBidder := &adapters.RequestData{
+		Method:  "POST",
+		Uri:     adapter.endpoint,
+		Body:    openRTBRequestJSON,
+		Headers: headers,
+	}
+	requestsToBidder = append(requestsToBidder, requestToBidder)
+
+	return requestsToBidder, errs
+}
+
+const unexpectedStatusCodeFormat = "" +
+	"Unexpected status code: %d. Run with request.debug = 1 for more info"
+
+// MakeBids unpacks the server's response into Bids.
+func (adapter *AdxcgAdapter) MakeBids(
+	openRTBRequest *openrtb2.BidRequest,
+	requestToBidder *adapters.RequestData,
+	bidderRawResponse *adapters.ResponseData,
+) (
+	bidderResponse *adapters.BidderResponse,
+	errs []error,
+) {
+	switch bidderRawResponse.StatusCode {
+	case http.StatusOK:
+		break
+	case http.StatusNoContent:
+		return nil, nil
+	case http.StatusBadRequest:
+		err := &errortypes.BadInput{
+			Message: fmt.Sprintf(unexpectedStatusCodeFormat, bidderRawResponse.StatusCode),
+		}
+		return nil, []error{err}
+	default:
+		err := &errortypes.BadServerResponse{
+			Message: fmt.Sprintf(unexpectedStatusCodeFormat, bidderRawResponse.StatusCode),
+		}
+		return nil, []error{err}
+	}
+
+	var openRTBBidderResponse openrtb2.BidResponse
+	if err := json.Unmarshal(bidderRawResponse.Body, &openRTBBidderResponse); err != nil {
+		return nil, []error{err}
+	}
+
+	bidsCapacity := len(openRTBBidderResponse.SeatBid[0].Bid)
+	bidderResponse = adapters.NewBidderResponseWithBidsCapacity(bidsCapacity)
+	var typedBid *adapters.TypedBid
+	for _, seatBid := range openRTBBidderResponse.SeatBid {
+		for _, bid := range seatBid.Bid {
+			bid := bid // pin! -> https://github.com/kyoh86/scopelint#whats-this
+			//TODO this must return data from extension
+
+			bidType, err := getMediaTypeForImp(bid.ImpID, openRTBRequest.Imp)
+			if err != nil {
+				errs = append(errs, err)
+				continue
+			}
+
+			typedBid = &adapters.TypedBid{Bid: &bid, BidType: bidType}
+			bidderResponse.Bids = append(bidderResponse.Bids, typedBid)
+		}
+	}
+
+	return bidderResponse, nil
+
+}
+
+func getMediaTypeForImp(impID string, imps []openrtb2.Imp) (openrtb_ext.BidType, error) {
+	for _, imp := range imps {
+		if imp.ID == impID {
+			if imp.Native != nil {
+				return openrtb_ext.BidTypeNative, nil
+			} else if imp.Banner != nil {
+				return openrtb_ext.BidTypeBanner, nil
+			} else if imp.Video != nil {
+				return openrtb_ext.BidTypeVideo, nil
+			}
```

Comment:
This is an anti-pattern. Please use the response from your bidding server to determine the bid type.

Reply:
Would like to ... but backend bidding servers don't have this property ready for Live enviornment.

Reply:
Gotcha. For now, please be sure the order in this code matches the order evaluated by your servers. For example, according to this code, native will always be bid on first. Incorrect bid type detection will cause rendering issues on the client.